pdns/recursor_cache.cc

```diff
@@ -152,7 +191,31 @@ time_t MemRecursorCache::handleHit(MapCombo::LockedContent& content, MemRecursor
   return ttd;
 }
 
-MemRecursorCache::cache_t::const_iterator MemRecursorCache::getEntryUsingECSIndex(MapCombo::LockedContent& map, time_t now, const DNSName& qname, const QType qtype, bool requireAuth, const ComboAddress& who)
+static void pushRefreshTask(const DNSName& qname, QType qtype, time_t deadline, const Netmask& netmask)
+{
+  if (qtype == QType::ADDR) {
+    pushAlmostExpiredTask(qname, QType::A, deadline, netmask);
+    pushAlmostExpiredTask(qname, QType::AAAA, deadline, netmask);
+  }
+  else {
+    pushAlmostExpiredTask(qname, qtype, deadline, netmask);
+  }
+}
+
+void MemRecursorCache::updateStaleEntry(time_t now, MemRecursorCache::OrderedTagIterator_t& entry)
+{
+  // We need to take care a infrequently access stale item cannot be extended past
+  // s_maxServedStaleExtension * s_serveStaleExtensionPeriod
+  // We we look how old the entry is, and increase d_servedStale accordingly, taking care not to overflow
+  const time_t howlong = std::max(static_cast<time_t>(1), now - entry->d_ttd);
+  const uint32_t extension = std::max(1U, std::min(entry->d_orig_ttl, s_serveStaleExtensionPeriod));
+  entry->d_servedStale = std::min(entry->d_servedStale + 1 + howlong / extension, static_cast<time_t>(s_maxServedStaleExtensions));
```

Comment:
So this is a bit more clever than `how many times the serve-stale status was extended`, since it takes into account for how long the entry was expired when we decide to use it as stale. Which we have to do not to exceed `s_maxServedStaleExtension * s_serveStaleExtensionPeriod`, given that we don't keep the original TTD around.
I see we check that `entry->d_servedStale < s_maxServedStaleExtensions` before entering this function, but since we can increment `d_servedStale` by more than one it looks like we could exceed `s_maxServedStaleExtensions` once?

Reply:
Line 202 assign `std::min(..., s_maxServedStaleExtension)` to `entry->d_servedStale` , so if the computed value is larger than `s_maxServedStaleExtension` the new value of `entry->d_servedStale` will be `s_maxServedStaleExtension`.

Reply:
Uh, right.